third_party/move/move-vm/runtime/src/storage/environment.rs

```diff
@@ -0,0 +1,315 @@
+    struct_name_index_map: StructNameIndexMap,
+
+    /// Type cache for struct layouts, tags and depths, shared across multiple threads.
+    ///
+    /// SAFETY:
+    /// Here we informally show that it is safe to share type cache across multiple threads. Same
+    /// argument applies to struct name indexing map.
+    ///   1) Struct has been already published.
+    ///      In this case, it is fine to have multiple transactions concurrently accessing and
+    ///      caching struct names, layouts and depth formulas. Even if transaction failed due to
+    ///      speculation, and is re-executed later, the speculative aborted execution cached a non-
+    ///      speculative existing struct information. It is safe for other threads to access it.
+    ///  2) Struct is being published with a module.
+    ///     The design of V2 loader ensures that when modules are published, i.e., staged on top of
+    ///     the existing module storage, the runtime environment is cloned. Hence, it is not even
+    ///     possible to mutate this global cache speculatively.
+    ty_cache: RwLock<TypeCache>,
+}
+
+impl RuntimeEnvironment {
+    /// Creates a new runtime environment with native functions and default VM configurations. If
+    /// there are duplicated natives, panics.
+    pub fn new(
+        natives: impl IntoIterator<Item = (AccountAddress, Identifier, Identifier, NativeFunction)>,
+    ) -> Self {
+        let vm_config = VMConfig {
+            // Keep the paranoid mode on as we most likely want this for tests.
+            paranoid_type_checks: true,
+            ..VMConfig::default()
+        };
+        Self::new_with_config(natives, vm_config)
+    }
+
+    /// Creates a new runtime environment with native functions and VM configurations. If there are
+    /// duplicated natives, panics.
+    pub fn new_with_config(
+        natives: impl IntoIterator<Item = (AccountAddress, Identifier, Identifier, NativeFunction)>,
+        vm_config: VMConfig,
+    ) -> Self {
+        let natives = NativeFunctions::new(natives)
+            .unwrap_or_else(|e| panic!("Failed to create native functions: {}", e));
+        Self {
+            vm_config,
+            natives,
+            verifier_extension: None,
+            struct_name_index_map: StructNameIndexMap::empty(),
+            ty_cache: RwLock::new(TypeCache::empty()),
+        }
+    }
+
+    /// Returns the config currently used by this runtime environment.
+    pub fn vm_config(&self) -> &VMConfig {
+        &self.vm_config
+    }
+
+    /// Enables delayed field optimization for this environment.
+    pub fn enable_delayed_field_optimization(&mut self) {
+        self.vm_config.delayed_field_optimization_enabled = true;
+    }
+
+    /// Creates a partially verified compiled script by running:
+    ///   1. Move bytecode verifier,
+    ///   2. Verifier extension, if provided.
+    pub fn build_partially_verified_script(
```

Comment:
This partial verification seems a bit vague to me and could be hard for us to maintain the code in the longer run. Does it mean modules that went through Move specific verification but not aptos specific ones? Ideally I would say we should not have sth partial but at least we need a clear definition of what's partial and what remains to be done.

Reply:
Good point, maybe the doctoring is not descriptive enough or naming can be better? Partial = local verification, which is based on the module. Full verification is also taking into account its dependencies, i.e. verifying imports.

Reply:
Old comment, but anyways resolved: called `locally` know